File: openfollow/web/static/js/units.js
/*
 * Client-side unit display/parse helpers (length conversion at input/display boundary).
 * Zone-editor canvas and camera wizard convert lengths; data model/API/scene math
 * stay in METRES. Faithful port of openfollow.units (format_length / parse_length /
 * metric_echo). Keep in sync with Python version (source of truth, has test suite).
 *
 * The active system is seeded from the server: base.tpl emits
 * ``window.OPENFOLLOW_UNIT_SYSTEM`` (from ``config.ui.unit_system``) before this
 * script loads. Callers use ``window.OpenFollow.units.{isImperial,formatLength,
 * parseLength,metricEcho,unitSuffixLength}``.
 */
(function () {
    window.OpenFollow = window.OpenFollow || {};

    // Exact SI definitions – 1 ft = 0.3048 m, 1 in = 0.0254 m.
    var M_PER_FT = 0.3048, M_PER_IN = 0.0254, IN_PER_FT = 12.0;
    var UNIT_TO_M = {m: 1.0, cm: 0.01, mm: 0.001, ft: M_PER_FT, "'": M_PER_FT, "in": M_PER_IN, '"': M_PER_IN};
    // Whole-magnitude validator: one-or-more unsigned tokens, nothing else.
    var LEN_FULL_RE = /^(?:\d*\.?\d+\s*(?:mm|cm|ft|in|m|'|")?\s*)+$/i;

    var imperial = false;

    function setSystem(sys) { imperial = (sys === 'imperial'); }
    function isImperial() { return imperial; }

    // Fixed-decimal string with round-half-to-even, matching Python's
    // format(x, '.Nf'). JS toFixed rounds half AWAY from zero, which diverges
    // from Python (banker's) for values that land exactly on a half (n/2^k like
    // 0.0625 m / 0.125 in) – and the "Stored: X.XXX m" echo is the operator's
    // contract for the persisted value, so it must agree with the server.
    function toFixedHalfEven(x, dp) {
        if (!isFinite(x)) return x.toFixed(dp);
        var neg = x < 0;
        var scaled = Math.abs(x) * Math.pow(10, dp);
        var floorVal = Math.floor(scaled);
        // ``scaled - floorVal === 0.5`` is exact only for dyadic x (where the
        // half is representable); every other value falls to nearest rounding,
        // matching Python's correctly-rounded format() on the same double.
        var n = (scaled - floorVal === 0.5)
            ? ((floorVal % 2 === 0) ? floorVal : floorVal + 1)
            : Math.round(scaled);
        var s = String(n);
        if (dp > 0) {
            while (s.length <= dp) s = '0' + s;
            s = s.slice(0, s.length - dp) + '.' + s.slice(s.length - dp);
        }
        return (neg ? '-' : '') + s;
    }

    // Metric: "1.524 m". Imperial adaptive: <1 ft -> "1.97 in";
    // >=1 ft -> "5 ft 4.85 in".
    function formatLength(meters) {
        if (!imperial) return toFixedHalfEven(meters, 3) + ' m';
        var sign = meters < 0 ? '-' : '';
        var a = Math.abs(meters);
        if (a < M_PER_FT) return sign + toFixedHalfEven(a / M_PER_IN, 2) + ' in';
        var totalIn = a / M_PER_IN;
        var ft = Math.floor(totalIn / IN_PER_FT);
        var remIn = totalIn - ft * IN_PER_FT;
        // 2-dp rounding can push the remainder to 12.00 in – carry it into the
        // feet column so we never render "4 ft 12.00 in" (matches Python).
        if (Math.round(remIn * 100) / 100 >= IN_PER_FT) { ft += 1; remIn = 0.0; }
        return sign + ft + ' ft ' + toFixedHalfEven(remIn, 2) + ' in';
    }

    // Always "X.XXX m" – the canonical stored value shown beside imperial inputs.
    function metricEcho(meters) { return toFixedHalfEven(meters, 3) + ' m'; }

    // Parse operator input to METRES; returns NaN on empty/garbage so callers
    // skip the update (a canvas only moves on a valid parse). Bare number =
    // feet in imperial / metres in metric; explicit m/cm/mm/ft/in/'/" work in
    // either mode; compounds like 5'6" / "5 ft 6 in"; one leading sign.
    function parseLength(raw) {
        var s = String(raw == null ? '' : raw).trim();
        if (!s) return NaN;
        var sign = 1.0;
        if (s[0] === '+' || s[0] === '-') { if (s[0] === '-') sign = -1.0; s = s.slice(1).trim(); }
        if (!LEN_FULL_RE.test(s)) return NaN;
        var defaultMult = imperial ? M_PER_FT : 1.0;
        // Local regex so there's no shared mutable ``lastIndex`` cursor across
        // calls. One token: a number with an optional unit suffix.
        var tokenRe = /(\d*\.?\d+)\s*(mm|cm|ft|in|m|'|")?/gi;
        var tokens = [], bare = 0, mt;
        while ((mt = tokenRe.exec(s)) !== null) {
            tokens.push(mt);
            if (!mt[2]) bare++;
        }
        if (!tokens.length) return NaN;
        // A bare number is a complete value; mixing it with units ("5 6 in") is
        // ambiguous – reject rather than guess (matches Python).
        if (bare && tokens.length > 1) return NaN;
        var total = 0.0;
        for (var i = 0; i < tokens.length; i++) {
            var unit = (tokens[i][2] || '').toLowerCase();
            total += parseFloat(tokens[i][1]) * (unit ? UNIT_TO_M[unit] : defaultMult);
        }
        var result = sign * total;
        return isFinite(result) ? result : NaN;
    }

    // Mirror openfollow.units.unit_suffix_length exactly ("m", not "meters").
    function unitSuffixLength() { return imperial ? 'ft / in' : 'm'; }

    window.OpenFollow.units = {
        setSystem: setSystem,
        isImperial: isImperial,
        formatLength: formatLength,
        parseLength: parseLength,
        metricEcho: metricEcho,
        unitSuffixLength: unitSuffixLength,
    };

    // base.tpl injected the active system before this script tag.
    if (window.OPENFOLLOW_UNIT_SYSTEM) setSystem(window.OPENFOLLOW_UNIT_SYSTEM);
})();

File: openfollow/units.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (C) 2026 OpenFollow Project
"""Display-unit conversion, formatting, and parsing.

Single source of truth for turning the internal **metric** model into
operator-facing strings and back. The data model, config storage, OSC,
PSN/RTTrPM/OTP, and scene math stay metric everywhere – only the web
UI and the on-device overlay route length / speed values through here.

Two unit systems, one global preference (``config.ui.unit_system``):

- ``metric``  – lengths in metres (3 dp), speed in m/s.
- ``imperial`` – lengths adaptive ft/in, speed in ft/s.

Imperial display is intentionally **lossy**: inches render to 2 dp
(≈0.01 in ≈ 0.25 mm of resolution), so ``parse(format(x))`` is only
guaranteed within ~0.0002 m of ``x``, not bit-exact. The web UI shows
a ``Stored: X.XXX m`` echo (``metric_echo`` of the *parsed* input) so
the operator always sees the exact value that will be persisted –
that's the contract, not a bit-exact round-trip.
"""

from __future__ import annotations

import math
import re
from enum import Enum


class UnitSystem(str, Enum):
    METRIC = "metric"
    IMPERIAL = "imperial"


# Exact SI conversion factors.
_M_PER_FT = 0.3048
_M_PER_IN = 0.0254
_FT_PER_M = 1.0 / _M_PER_FT
_IN_PER_FT = 12.0

# Unit token to metres conversion. Bare '' resolves to mode default at parse time.
_UNIT_TO_METERS: dict[str, float] = {
    "m": 1.0,
    "cm": 0.01,
    "mm": 0.001,
    "ft": _M_PER_FT,
    "'": _M_PER_FT,
    "in": _M_PER_IN,
    '"': _M_PER_IN,
}

# Parse length tokens; validate complete input.
_LEN_TOKEN_RE = re.compile(r"(\d*\.?\d+)\s*(mm|cm|ft|in|m|'|\")?", re.IGNORECASE)
_LEN_FULL_RE = re.compile(r"(?:\d*\.?\d+\s*(?:mm|cm|ft|in|m|'|\")?\s*)+", re.IGNORECASE)

_SPEED_RE = re.compile(r"([+-]?\d*\.?\d+)\s*(m/s|ft/s)?", re.IGNORECASE)


# --------------------------------------------------------------------------- #
# Formatting
# --------------------------------------------------------------------------- #


def format_length(meters: float, unit_system: UnitSystem) -> str:
    """Format metric length for display: ``"X.XXX m"`` or ``"X ft Y.YY in"``."""
    if unit_system is UnitSystem.METRIC:
        return f"{meters:.3f} m"
    sign = "-" if meters < 0 else ""
    a = abs(meters)
    total_in = a / _M_PER_IN
    # Inches-only form only when it won't round up to a full foot: a value
    # just under 1 ft (e.g. 11.998 in) would otherwise print "12.00 in", a
    # nonsensical imperial form – carry to "1 ft 0.00 in" instead.
    if a < _M_PER_FT and round(total_in, 2) < _IN_PER_FT:
        return f"{sign}{total_in:.2f} in"
    ft = int(total_in // _IN_PER_FT)
    rem_in = total_in - ft * _IN_PER_FT
    # Rounding can overflow to 12.00 in; carry to feet.
    if round(rem_in, 2) >= _IN_PER_FT:
        ft += 1
        rem_in = 0.0
    return f"{sign}{ft} ft {rem_in:.2f} in"


def format_length_compact(meters: float, unit_system: UnitSystem) -> str:
    """HUD compact length: fixed-width signed number, no unit suffix."""
    if unit_system is UnitSystem.METRIC:
        return f"{meters:+6.2f}"
    return f"{meters * _FT_PER_M:+6.2f}"


def format_speed(mps: float, unit_system: UnitSystem) -> str:
    """Format metric speed: ``"X.XX m/s"`` or ``"X.XX ft/s"``."""
    if unit_system is UnitSystem.METRIC:
        return f"{mps:.2f} m/s"
    return f"{mps * _FT_PER_M:.2f} ft/s"


def metric_echo(meters: float) -> str:
    """Return metric representation ``"X.XXX m"`` for UI echo of stored value."""
    return f"{meters:.3f} m"


def metric_echo_speed(mps: float) -> str:
    """Return metric speed representation ``"X.XXX m/s"`` for UI echo."""
    return f"{mps:.3f} m/s"


def unit_suffix_length(unit_system: UnitSystem) -> str:
    return "m" if unit_system is UnitSystem.METRIC else "ft / in"


def unit_suffix_speed(unit_system: UnitSystem) -> str:
    return "m/s" if unit_system is UnitSystem.METRIC else "ft/s"


# --------------------------------------------------------------------------- #
# Parsing
# --------------------------------------------------------------------------- #


def parse_length(raw: str, unit_system: UnitSystem) -> float:
    """Parse operator input into metres: supports numbers, units (m/cm/mm/ft/in), and compound forms."""
    s = raw.strip()
    if not s:
        raise ValueError("empty length")
    sign = 1.0
    if s[0] in "+-":
        if s[0] == "-":
            sign = -1.0
        s = s[1:].strip()
    if not _LEN_FULL_RE.fullmatch(s):
        raise ValueError(f"unparseable length: {raw!r}")
    default_mult = _M_PER_FT if unit_system is UnitSystem.IMPERIAL else 1.0
    # ``_LEN_FULL_RE`` (one-or-more tokens) already matched, so finditer
    # yields at least one token.
    tokens = list(_LEN_TOKEN_RE.finditer(s))
    bare = [t for t in tokens if not t.group(2)]
    # Reject ambiguous input like "5 6 in" (bare number mixed with units).
    if bare and len(tokens) > 1:
        raise ValueError(f"ambiguous length (bare number with units): {raw!r}")
    total = 0.0
    for t in tokens:
        value = float(t.group(1))
        unit = (t.group(2) or "").lower()
        mult = _UNIT_TO_METERS[unit] if unit else default_mult
        total += value * mult
    result = sign * total
    if not math.isfinite(result):
        raise ValueError(f"length out of range: {raw!r}")
    return result


def parse_speed(raw: str, unit_system: UnitSystem) -> float:
    """Parse operator input into m/s; accepts m/s and ft/s explicitly, or bare number in mode default."""
    s = raw.strip()
    if not s:
        raise ValueError("empty speed")
    m = _SPEED_RE.fullmatch(s)
    if m is None:
        raise ValueError(f"unparseable speed: {raw!r}")
    value = float(m.group(1))
    if not math.isfinite(value):
        raise ValueError(f"speed out of range: {raw!r}")
    unit = (m.group(2) or "").lower()
    if unit == "m/s":
        return value
    if unit == "ft/s":
        return value * _M_PER_FT
    # Bare number → mode default.
    return value * (_M_PER_FT if unit_system is UnitSystem.IMPERIAL else 1.0)
